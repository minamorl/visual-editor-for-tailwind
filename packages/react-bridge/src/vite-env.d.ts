/// <reference types="vite/client" />

declare global {
  interface Window {
    __REACT_DEVTOOLS_GLOBAL_HOOK__: any;
  }
}
