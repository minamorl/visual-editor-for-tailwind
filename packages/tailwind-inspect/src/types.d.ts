namespace JSX {
  interface IntrinsicElements {
    ["iconify-icon"]: React.DetailedHTMLProps<
      React.HTMLAttributes<HTMLElement>,
      HTMLElement
    > & {
      class?: string;
      icon?: string;
    };
  }
}
