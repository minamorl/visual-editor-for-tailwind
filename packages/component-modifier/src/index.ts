import { parse } from "@babel/parser";
import generate from "@babel/generator";
import traverse from "@babel/traverse";

export const getAttributeModifiedCode = (arg: {
  code: string;
  targetComponentLine: number;
  attribute: string;
  value: string;
}) => {
  const { code, targetComponentLine, attribute, value } = arg;

  const ast = parse(code, {
    sourceType: "module",
    plugins: ["jsx", "typescript"],
  });

  traverse(ast, {
    enter(path) {
      const isUpdate =
        path.isJSXAttribute({ name: attribute }) &&
        path.parent.loc?.start.line === targetComponentLine;

      const isAdd =
        path.isJSXOpeningElement() &&
        path.node.loc?.start.line === targetComponentLine;

      if (isUpdate) {
        path.node.value = { type: "StringLiteral", value };
        return;
      }
      if (isAdd) {
        path.node.attributes.unshift({
          type: "JSXAttribute",
          name: { type: "JSXIdentifier", name: attribute },
          value: { type: "StringLiteral", value },
        });
        return;
      }
      // TODO remove
    },
  });

  return generate(ast).code;
};
