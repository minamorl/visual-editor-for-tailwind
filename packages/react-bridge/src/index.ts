import {
  activate as activateBackend,
  initialize as initializeBackend,
} from "react-devtools-inline/backend";
import { initialize as initializeFrontend } from "react-devtools-inline/frontend";

initializeBackend(window);
initializeFrontend(window);

const initBridge = () => {
  activateBackend(window);

  let tmpBorder = "";
  window.addEventListener("pointerout", (e) => {
    if (!e.target || !window.__REACT_DEVTOOLS_GLOBAL_HOOK__) return;

    const target = e.target as HTMLElement;
    target.style.border = tmpBorder;
  });

  window.addEventListener("pointerover", (e) => {
    if (!e.target || !window.__REACT_DEVTOOLS_GLOBAL_HOOK__) return;

    const renderer = window.__REACT_DEVTOOLS_GLOBAL_HOOK__.renderers.get(1);
    if (!renderer) {
      console.error("No __REACT_DEVTOOLS_GLOBAL_HOOK__ renderer !!");
      return;
    }

    const target = e.target as HTMLElement;
    tmpBorder = target.style.border;

    const fiber = renderer.findFiberByHostInstance(e.target);

    if (fiber && fiber._debugSource) {
      target.title = `${fiber._debugSource.fileName}:${fiber._debugSource.lineNumber}`;
      target.style.border = "1px solid blue";

      window.parent.postMessage(
        { source: "tailwind-editor", _debugSource: fiber._debugSource },
        "*"
      );
      console.log(fiber._debugSource);
    }
  });
};

window.addEventListener("load", initBridge);
