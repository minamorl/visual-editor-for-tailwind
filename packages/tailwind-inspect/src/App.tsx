import { PaintkitRoot } from "@seanchas116/paintkit/src/components/PaintkitRoot";
import { action } from "mobx";
import { observer } from "mobx-react-lite";
import React from "react";
import { useEffect, useRef } from "react";
import { StyleInspector } from "./inspector/StyleInspector";
import { JSXTreeView } from "./outline/JSXTreeView";
import { AppState } from "./state/AppState";

const appState = new AppState();

const App = observer(function App() {
  const tailwindClass = appState.elementInstance.style.toTailwind();
  const targetRef = useRef<HTMLDivElement>(null);

  useEffect(
    action(() => {
      if (targetRef.current) {
        appState.elementInstance.computedStyle.loadComputedStyle(
          targetRef.current
        );
      }
    }),
    [tailwindClass]
  );

  return (
    <PaintkitRoot colorScheme="dark">
      <div className="flex w-full h-full fixed left-0 top-0">
        <div className="flex-1">
          <div className={tailwindClass} ref={targetRef}>
            Lorem Ipsum
            <div className="bg-red-500 w-24 h-12">Block 1</div>
            <div className="bg-blue-500 w-12 h-24">Block 2</div>
          </div>
          <div className="text-gray-500 text-sm">
            class: <span className="text-gray-700">{tailwindClass}</span>
          </div>
          <DemoRunner appState={appState} />
        </div>
        <div className="bg-zinc-800 w-64 flex flex-col ">
          <JSXTreeView
            className="h-80 shrink-0 border-b-neutral-700 border-solid border-b-[2px]"
            file={appState.sourceFile}
          />
          <StyleInspector
            className="flex-1 overflow-y-auto"
            state={appState.styleInspectorState}
          />
        </div>
      </div>
    </PaintkitRoot>
  );
});

const DemoRunner = observer(({ appState }: { appState: AppState }) => {
  const module = new Function("exports", "React", appState.compiledCode) as (
    exports: any,
    react: typeof React
  ) => void;
  const exports: { default?: React.FC } = {};
  module(exports, React);

  const Component = exports.default!;

  return <Component />;
});

export default App;
