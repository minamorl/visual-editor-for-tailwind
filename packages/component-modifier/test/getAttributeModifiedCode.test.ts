import { test, describe, expect } from "vitest";
import { readFileSync } from "fs";
import { getAttributeModifiedCode } from "../src";
import path from "path";

describe("getAttributeModifiedCode", () => {
  const getTestCodeByFileName = (fileName: string) =>
    readFileSync(
      path.join(__dirname, "./data/getAttributeModifiedCode", fileName),
      "utf8"
    );

  const testCases = [
    {
      name: "jsx, update one line component",
      inputCode: getTestCodeByFileName("Card.jsx"),
      targetComponentLine: 7,
      attribute: "className",
      value: "mod text-xl mb-2",
    },
    {
      name: "jsx, update multi line component",
      inputCode: getTestCodeByFileName("Card.jsx"),
      targetComponentLine: 8,
      attribute: "className",
      value: "mod text-base",
    },
    {
      name: "tsx, update one line component",
      inputCode: getTestCodeByFileName("Card.tsx"),
      targetComponentLine: 16,
      attribute: "className",
      value: "mod text-xl mb-2",
    },
    {
      name: "tsx, update multi line component",
      inputCode: getTestCodeByFileName("Card.tsx"),
      targetComponentLine: 17,
      attribute: "className",
      value: "mod text-base",
    },
    {
      name: "jsx, add one line component",
      inputCode: getTestCodeByFileName("NoClassNameComponent.jsx"),
      targetComponentLine: 5,
      attribute: "className",
      value: "mod",
    },
    {
      name: "jsx, add multi line component",
      inputCode: getTestCodeByFileName("NoClassNameComponent.jsx"),
      targetComponentLine: 6,
      attribute: "className",
      value: "mod",
    },
    {
      name: "jsx, remove one line component",
      inputCode: getTestCodeByFileName("Card.jsx"),
      targetComponentLine: 4,
      attribute: "className",
      value: "",
    },
    {
      name: "jsx, remove multi line component",
      inputCode: getTestCodeByFileName("Card.jsx"),
      targetComponentLine: 8,
      attribute: "className",
      value: "",
    },
  ];

  testCases.forEach(
    ({ name, inputCode: code, targetComponentLine, attribute, value }) => {
      test(name, () => {
        const arg = { code, targetComponentLine, attribute, value };
        const result = getAttributeModifiedCode(arg);
        expect(result).toMatchSnapshot();
      });
    }
  );
});
