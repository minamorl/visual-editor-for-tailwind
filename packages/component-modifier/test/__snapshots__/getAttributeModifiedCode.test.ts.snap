// Vitest Snapshot v1

exports[`getAttributeModifiedCode > jsx, multi line component 1`] = `
"import React from \\"react\\";

const Card = ({
  title,
  image,
  onClick
}) => <div className=\\"max-w-sm rounded overflow-hidden shadow-lg\\" onClick={onClick}>
    <img className=\\"w-full\\" src={image} alt=\\"Sunset in the mountains\\" />
    <div className=\\"px-6 py-4\\">
      <div className=\\"font-bold text-xl mb-2\\">{title}</div>
      <p className=\\"mod text-base\\" style={{
      alignItems: \\"center\\",
      textAlign: \\"center\\"
    }} className=\\"text-gray-700 text-base\\">
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
        quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
        nihil.
      </p>
    </div>
    <div className=\\"px-6 pt-4 pb-2\\">
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #photography
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #travel
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #winter
      </span>
    </div>
  </div>;

export default Card;"
`;

exports[`getAttributeModifiedCode > jsx, no className multi line component 1`] = `
"import React from \\"react\\";

const NoClassNameComponent = () => <div>
    <div>title</div>
    <p className=\\"mod\\" data-1=\\"1\\" data-2=\\"2\\" data-3=\\"3\\" data-4=\\"4\\" data-5=\\"5\\" data-6=\\"6\\" data-7=\\"7\\" data-8=\\"8\\" data-9=\\"9\\" data-10=\\"10\\">
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
      quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
      nihil.
    </p>
  </div>;

export default NoClassNameComponent;"
`;

exports[`getAttributeModifiedCode > jsx, no className one line component 1`] = `
"import React from \\"react\\";

const NoClassNameComponent = () => <div>
    <div className=\\"mod\\">title</div>
    <p data-1=\\"1\\" data-2=\\"2\\" data-3=\\"3\\" data-4=\\"4\\" data-5=\\"5\\" data-6=\\"6\\" data-7=\\"7\\" data-8=\\"8\\" data-9=\\"9\\" data-10=\\"10\\">
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
      quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
      nihil.
    </p>
  </div>;

export default NoClassNameComponent;"
`;

exports[`getAttributeModifiedCode > jsx, one line component 1`] = `
"import React from \\"react\\";

const Card = ({
  title,
  image,
  onClick
}) => <div className=\\"max-w-sm rounded overflow-hidden shadow-lg\\" onClick={onClick}>
    <img className=\\"w-full\\" src={image} alt=\\"Sunset in the mountains\\" />
    <div className=\\"px-6 py-4\\">
      <div className=\\"mod text-xl mb-2\\" className=\\"font-bold text-xl mb-2\\">{title}</div>
      <p style={{
      alignItems: \\"center\\",
      textAlign: \\"center\\"
    }} className=\\"text-gray-700 text-base\\">
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
        quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
        nihil.
      </p>
    </div>
    <div className=\\"px-6 pt-4 pb-2\\">
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #photography
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #travel
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #winter
      </span>
    </div>
  </div>;

export default Card;"
`;

exports[`getAttributeModifiedCode > tsx, multi line component 1`] = `
"// @ts-nocheck
import React from \\"react\\";

const Card = ({
  title,
  image,
  onClick
}: {
  title: string;
  image: string;
  onClick: () => void;
}) => <div className=\\"max-w-sm rounded overflow-hidden shadow-lg\\" onClick={onClick}>
    <img className=\\"w-full\\" src={image} alt=\\"Sunset in the mountains\\" />
    <div className=\\"px-6 py-4\\">
      <div className=\\"font-bold text-xl mb-2\\">{title}</div>
      <p className=\\"mod text-base\\" style={{
      alignItems: \\"center\\",
      textAlign: \\"center\\"
    }} className=\\"text-gray-700 text-base\\">
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
        quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
        nihil.
      </p>
    </div>
    <div className=\\"px-6 pt-4 pb-2\\">
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #photography
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #travel
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #winter
      </span>
    </div>
  </div>;

export default Card;"
`;

exports[`getAttributeModifiedCode > tsx, one line component 1`] = `
"// @ts-nocheck
import React from \\"react\\";

const Card = ({
  title,
  image,
  onClick
}: {
  title: string;
  image: string;
  onClick: () => void;
}) => <div className=\\"max-w-sm rounded overflow-hidden shadow-lg\\" onClick={onClick}>
    <img className=\\"w-full\\" src={image} alt=\\"Sunset in the mountains\\" />
    <div className=\\"px-6 py-4\\">
      <div className=\\"mod text-xl mb-2\\" className=\\"font-bold text-xl mb-2\\">{title}</div>
      <p style={{
      alignItems: \\"center\\",
      textAlign: \\"center\\"
    }} className=\\"text-gray-700 text-base\\">
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
        quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
        nihil.
      </p>
    </div>
    <div className=\\"px-6 pt-4 pb-2\\">
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #photography
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #travel
      </span>
      <span className=\\"inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2\\">
        #winter
      </span>
    </div>
  </div>;

export default Card;"
`;
